zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/SaslServerCallbackHandler.java

```diff
@@ -90,6 +90,11 @@ private void handleNameCallback(NameCallback nc) {
         // check to see if this user is in the user password database.
         if (credentials.get(nc.getDefaultName()) == null) {
             LOG.warn("User '{}' not found in list of DIGEST-MD5 authenticateable users.", nc.getDefaultName());
+            // ZOOKEEPER-4839
+            // Incorrect usernames also need to be stored
+            // in order to clear the usernames of previously
+            // successfully logged-in users.
+            userName = nc.getDefaultName();
```

Comment:
I saw one `SaslServerCallbackHandler` was shared among multiple concurrent `ServerCnxn`s. I think we probably should resort to some form of copy/clone to not share mutable state among multiple `ServerCnxn`s.

This applies to `SaslQuorumServerCallbackHandler` too.

Reply:
Thank you for your review. I have some ideas about your comments.

1. It would be better for the `Login` to store a factory object that can supply `SaslServerCallbackHandler`, in order to avoid the `SaslServerCallbackHandler` being shared among multiple concurrent `ServerCnxn`s.
2. We should apply the change to the `SaslQuorumServerCallbackHandler`.

Is it correct?

Reply:
Sounds viable to me.